### DISCUSSION TITLE
Breaking changes due for VcRedist 4.1

### DISCUSSION THREAD

**Discussion by aaronparker** [EUCPilots/vcredist#123] [General]
VcRedist `4.0` will provide mostly under-the-hood code improvements; however, VcRedist `4.1` will introduce two breaking changes:

* Functions including `Install-VcRedist`, `New-VcMdtApplication`, `New-VcConfigMgrApplication`, `New-VcIntuneApplication` etc., will require output from `Save-VcRedist` instead of `Get-VcList`. This will be done to simplify determining the paths to the downloaded vcredist.exe installers - `Get-VcList | Save-VcRedist` returns a `Path` property which means that a `-Path` parameter will not be required by other functions. **This change will require scripts to be updated**
* VcRedist 2005 and 2008 releases that have been pulled for download by Microsoft will be removed from the manifest. See [#114](https://github.com/aaronparker/vcredist/issues/114). This should be a minor issue anyway as these downloads don't work right now and the later 2005 and 2008 installers should be used. Note that 2005, 2008 and 2010 redistributables are no longer supported by MS

**Comment by aaronparker** [comment]
Here's some examples of the syntax change change - here's the change in syntax for `Install-VcRedist`

**Old syntax**

```powershell
$VcList = Get-VcList
Save-VcRedist -VcList $VcList -Path C:\Temp\VcRedist
Install-VcRedist -VcList $VcList -Path C:\Temp\VcRedist
```

**New syntax**

```powershell
Get-VcList | Save-VcRedist -Path C:\Temp\VcRedist | Install-VcRedist
```

**Comment by aaronparker** [comment]
Here's the change in syntax for `Import-VcMdtApplication`

**Old syntax**

```powershell
$VcList = Get-VcList
Save-VcRedist -VcList $VcList -Path C:\Temp\VcRedist
Import-VcMdtApplication -VcList $VcList -Path C:\Temp\VcRedist -MdtPath \\server\deployment
```

**New syntax**

```powershell
Get-VcList | Save-VcRedist -Path C:\Temp\VcRedist | Import-VcMdtApplication -MdtPath \\server\deployment
```

**Comment by aaronparker** [comment]
Here's the change in syntax for `Import-VcIntuneApplication`

**Old syntax**

```powershell
$VcList = Get-VcList
Save-VcRedist -VcList $VcList -Path C:\Temp\VcRedist
Import-VcIntuneApplication -VcList $VcList
```

**New syntax**

```powershell
Get-VcList | Save-VcRedist -Path C:\Temp\VcRedist | Import-VcIntuneApplication
```

**Reply by alexhass** [reply]
This ends with

```
Add-IntuneWin32App : Cannot validate argument on parameter 'FilePath'. Cannot bind argument to parameter 'Path' because it is an empty string.
At C:\Program Files\WindowsPowerShell\Modules\VcRedist\4.1.495\Public\Import-VcIntuneApplication.ps1:115 char:47
+             $Application = Add-IntuneWin32App @Win32AppArgs
+                                               ~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Add-IntuneWin32App], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationError,Add-IntuneWin32App
```

**Comment by aaronparker** [comment]
Here's the change in syntax for `Import-VcConfigMgrApplication`

**Old syntax**

```powershell
$VcList = Get-VcList
Save-VcRedist -VcList $VcList -Path C:\Temp\VcRedist
Import-VcConfigMgrApplication -VcList $VcList -Path C:\Temp\VcRedist -CMPath \\server\applications -SMSSiteCode LAB
```

**New syntax**

```powershell
Get-VcList | Save-VcRedist -Path C:\Temp\VcRedist | Import-VcConfigMgrApplication -CMPath \\server\applications -SMSSiteCode LAB
```

**Comment by aaronparker** [comment]
Note - I'll push these changes to the PowerShell Gallery on March 1, 2023.

**Reply by bmitchell76** [reply]
@aaronparker, did you push the update?  I'm still pulling v4.0.461 from the gallery.

**Comment by aaronparker** [comment]
ah yeah no. I got busy and completely forgot. Let's push this back

**Comment by aaronparker** [comment]
I'll push the updated module to the PowerShell gallery on March 15th. AEST. (i.e. March 14th for those in the US)

**Comment by aaronparker** [comment]
4.1 has now been pushed to the PowerShell Gallery.

**Reply by bmitchell76** [reply]
Thanks!  Testing it now, was proxy support dropped?

**Reply by aaronparker** [reply]
I don't think so. Mind you, I'm not actively testing proxy support.

**Reply by bmitchell76** [reply]
Well, before v4.1, we could pass -proxy and -proxycredentials.  I was able to work around it.

**Reply by bmitchell76** [reply]
We were supposed to have a significant proxy server change where we couldn't use basic auth, but that was postponed, so my workaround will work for now.

I appreciate your work!

**Reply by nahidf** [reply]
We are having issue installing the power shell module on docker container, its pointing to new version now. is there any change on quick install not mentioned in docs?

**Reply by aaronparker** [reply]
@bmitchell76 I'm not sure what's going on there - could you open an issue?

**Reply by aaronparker** [reply]
@nahidf are you using `https://vcredist.com/install.ps1`?

I've pushed a new version of that script, so you should see it update soon.

**Reply by nahidf** [reply]
@aaronparker Yes and I confirm its fixed now, thanks!

**Comment by God-damnit-all** [comment]
Not really too relevant since you can do `Get-VcList -Export All` or `Get-VcList -Export Unsupported`, but I would like to point out that while Microsoft has dropped support for those redistributables, they still make them available through [winget](https://github.com/microsoft/winget-pkgs/tree/master/manifests/m/Microsoft/VCRedist/2005).

If you ever decide to support more than just the English locale, their manifests are an [excellent resource](https://github.com/microsoft/winget-pkgs/blob/master/manifests/m/Microsoft/VCRedist/2005/x64/8.0.61000/Microsoft.VCRedist.2005.x64.installer.yaml#L38). You could even use GitHub Actions to auto-update `VisualCRedistributables.json` using these manifests.

**Comment by ei-zascha** [comment]
Doesn't this force you to download the redistributables every single time though? What if I want to download them once to a central location and install to several computers from there?

Also, it'd be helpful if the documentation at https://vcredist.com/ was updated to reflect these changes.

Thanks!

**Comment by aaronparker** [comment]
@aizascha if the vcredists are already downloaded, they won't be redownloaded.
